### ISSUE TITLE
AssertionError: can only join a child process

### ISSUE THREAD

**Issue by jd** [sileht/cotyledon#4]
When starting gnocchi-metricd using `pifpaf`. I think it's a bug in Cotyledon, but I'm not sure :(

```
DEBUG: pifpaf.drivers: executing: ['gnocchi-metricd', '--config-file=/var/folders/7k/pwdhb_mj2cv4zyr0kyrlzjx40000gq/T/tmp_heguw/gnocchi.conf']
DEBUG: pifpaf.drivers: gnocchi-metricd[46224] output: 2016-08-02 18:11:16.633 46241 CRITICAL gnocchi [-] IOError: [Errno 22] Invalid argument
DEBUG: pifpaf.drivers: gnocchi-metricd[46224] output: 2016-08-02 18:11:16.633 46241 ERROR gnocchi Traceback (most recent call last):
DEBUG: pifpaf.drivers: gnocchi-metricd[46224] output: 2016-08-02 18:11:16.633 46241 ERROR gnocchi   File "/usr/local/bin/gnocchi-metricd", line 10, in <module>
DEBUG: pifpaf.drivers: gnocchi-metricd[46224] output: 2016-08-02 18:11:16.633 46241 ERROR gnocchi     sys.exit(metricd())
DEBUG: pifpaf.drivers: gnocchi-metricd[46224] output: 2016-08-02 18:11:16.633 46241 ERROR gnocchi   File "/Users/jd/Source/gnocchi/gnocchi/cli.py", line 332, in metricd
DEBUG: pifpaf.drivers: gnocchi-metricd[46224] output: 2016-08-02 18:11:16.633 46241 ERROR gnocchi     MetricdServiceManager(conf).run()
DEBUG: pifpaf.drivers: gnocchi-metricd[46224] output: 2016-08-02 18:11:16.633 46241 ERROR gnocchi   File "/usr/local/lib/python2.7/site-packages/cotyledon/__init__.py", line 273, in run
DEBUG: pifpaf.drivers: gnocchi-metricd[46224] output: 2016-08-02 18:11:16.633 46241 ERROR gnocchi     pid = self._start_service(conf, worker_id)
DEBUG: pifpaf.drivers: gnocchi-metricd[46224] output: 2016-08-02 18:11:16.633 46241 ERROR gnocchi   File "/usr/local/lib/python2.7/site-packages/cotyledon/__init__.py", line 426, in _start_service
DEBUG: pifpaf.drivers: gnocchi-metricd[46224] output: 2016-08-02 18:11:16.633 46241 ERROR gnocchi     time.sleep(1000000000)
DEBUG: pifpaf.drivers: gnocchi-metricd[46224] output: 2016-08-02 18:11:16.633 46241 ERROR gnocchi IOError: [Errno 22] Invalid argument
DEBUG: pifpaf.drivers: gnocchi-metricd[46224] output: 2016-08-02 18:11:16.633 46241 ERROR gnocchi
DEBUG: pifpaf.drivers: gnocchi-metricd[46224] output: Error in atexit._run_exitfuncs:
DEBUG: pifpaf.drivers: gnocchi-metricd[46224] output: Traceback (most recent call last):
DEBUG: pifpaf.drivers: gnocchi-metricd[46224] output:   File "/usr/local/Cellar/python/2.7.12/Frameworks/Python.framework/Versions/2.7/lib/python2.7/atexit.py", line 24, in _run_exitfuncs
DEBUG: pifpaf.drivers: gnocchi-metricd[46224] output:     func(*targs, **kargs)
DEBUG: pifpaf.drivers: gnocchi-metricd[46224] output:   File "/usr/local/Cellar/python/2.7.12/Frameworks/Python.framework/Versions/2.7/lib/python2.7/multiprocessing/util.py", line 325, in _exit_function
DEBUG: pifpaf.drivers: gnocchi-metricd[46224] output:     p.join()
DEBUG: pifpaf.drivers: gnocchi-metricd[46224] output:   File "/usr/local/Cellar/python/2.7.12/Frameworks/Python.framework/Versions/2.7/lib/python2.7/multiprocessing/process.py", line 143, in join
DEBUG: pifpaf.drivers: gnocchi-metricd[46224] output: 2016-08-02 18:11:16.630 46239 CRITICAL gnocchi [-] IOError: [Errno 22] Invalid argument
DEBUG: pifpaf.drivers: gnocchi-metricd[46224] output: 2016-08-02 18:11:16.630 46239 ERROR gnocchi Traceback (most recent call last):
DEBUG: pifpaf.drivers: gnocchi-metricd[46224] output: 2016-08-02 18:11:16.630 46239 ERROR gnocchi   File "/usr/local/bin/gnocchi-metricd", line 10, in <module>
DEBUG: pifpaf.drivers: gnocchi-metricd[46224] output: 2016-08-02 18:11:16.630 46239 ERROR gnocchi     sys.exit(metricd())
DEBUG: pifpaf.drivers: gnocchi-metricd[46224] output: 2016-08-02 18:11:16.630 46239 ERROR gnocchi   File "/Users/jd/Source/gnocchi/gnocchi/cli.py", line 332, in metricd
DEBUG: pifpaf.drivers: gnocchi-metricd[46224] output: 2016-08-02 18:11:16.630 46239 ERROR gnocchi     MetricdServiceManager(conf).run()
DEBUG: pifpaf.drivers: gnocchi-metricd[46224] output: 2016-08-02 18:11:16.630 46239 ERROR gnocchi   File "/usr/local/lib/python2.7/site-packages/cotyledon/__init__.py", line 273, in run
DEBUG: pifpaf.drivers: gnocchi-metricd[46224] output: 2016-08-02 18:11:16.630 46239 ERROR gnocchi     pid = self._start_service(conf, worker_id)
DEBUG: pifpaf.drivers: gnocchi-metricd[46224] output: 2016-08-02 18:11:16.630 46239 ERROR gnocchi   File "/usr/local/lib/python2.7/site-packages/cotyledon/__init__.py", line 426, in _start_service
DEBUG: pifpaf.drivers: gnocchi-metricd[46224] output: 2016-08-02 18:11:16.630 46239 ERROR gnocchi     time.sleep(1000000000)
DEBUG: pifpaf.drivers: gnocchi-metricd[46224] output: 2016-08-02 18:11:16.630 46239 ERROR gnocchi IOError: [Errno 22] Invalid argument
DEBUG: pifpaf.drivers: gnocchi-metricd[46224] output: 2016-08-02 18:11:16.630 46239 ERROR gnocchi
DEBUG: pifpaf.drivers: gnocchi-metricd[46224] output:     assert self._parent_pid == os.getpid(), 'can only join a child process'
DEBUG: pifpaf.drivers: gnocchi-metricd[46224] output: AssertionError: can only join a child process
DEBUG: pifpaf.drivers: gnocchi-metricd[46224] output: Error in sys.exitfunc:
DEBUG: pifpaf.drivers: gnocchi-metricd[46224] output: Error in atexit._run_exitfuncs:
DEBUG: pifpaf.drivers: gnocchi-metricd[46224] output: Traceback (most recent call last):
DEBUG: pifpaf.drivers: gnocchi-metricd[46224] output:   File "/usr/local/Cellar/python/2.7.12/Frameworks/Python.framework/Versions/2.7/lib/python2.7/atexit.py", line 24, in _run_exitfuncs
DEBUG: pifpaf.drivers: gnocchi-metricd[46224] output:     func(*targs, **kargs)
DEBUG: pifpaf.drivers: gnocchi-metricd[46224] output:   File "/usr/local/Cellar/python/2.7.12/Frameworks/Python.framework/Versions/2.7/lib/python2.7/multiprocessing/util.py", line 325, in _exit_function
DEBUG: pifpaf.drivers: gnocchi-metricd[46224] output:     p.join()
DEBUG: pifpaf.drivers: gnocchi-metricd[46224] output:   File "/usr/local/Cellar/python/2.7.12/Frameworks/Python.framework/Versions/2.7/lib/python2.7/multiprocessing/process.py", line 143, in join
DEBUG: pifpaf.drivers: gnocchi-metricd[46224] output: 2016-08-02 18:11:16.683 46241 CRITICAL gnocchi [-] AssertionError: can only join a child process
```

**Comment by sileht** [comment]
I have released 1.2.7 that fixes this issue. Now the backtrace show two issue on in Gnocchi and one in Pifpaf.

I have fixed the pifpaf one: https://github.com/jd/pifpaf/pull/18
I have not investigated the Gnocchi one

**Comment by sileht** [comment]
The new backtrace:

DEBUG: pifpaf.drivers: gnocchi-metricd[28413] output: 2016-08-03 00:51:50.229 28444 INFO cotyledon [-] Caught SIGTERM signal, graceful exiting of service processing(2) [28444]
DEBUG: pifpaf.drivers: gnocchi-metricd[28413] output: 2016-08-03 00:51:50.229 28441 INFO cotyledon [-] Caught SIGTERM signal, graceful exiting of service processing(1) [28441]
DEBUG: pifpaf.drivers: gnocchi-metricd[28413] output: 2016-08-03 00:51:50.229 28444 ERROR cotyledon [-] Unhandled exception
DEBUG: pifpaf.drivers: gnocchi-metricd[28413] output: 2016-08-03 00:51:50.229 28444 ERROR cotyledon Traceback (most recent call last):
DEBUG: pifpaf.drivers: gnocchi-metricd[28413] output: 2016-08-03 00:51:50.229 28444 ERROR cotyledon   File "/workspace/sileht/os_dev/stack/python-gnocchiclient/.tox/py27/local/lib/python2.7/site-packages/cotyledon/__init__.py", line 52, in _exit_on_exception
DEBUG: pifpaf.drivers: gnocchi-metricd[28413] output: 2016-08-03 00:51:50.229 28444 ERROR cotyledon     yield
DEBUG: pifpaf.drivers: gnocchi-metricd[28413] output: 2016-08-03 00:51:50.229 28444 ERROR cotyledon   File "/workspace/sileht/os_dev/stack/python-gnocchiclient/.tox/py27/local/lib/python2.7/site-packages/cotyledon/__init__.py", line 141, in _clean_exit
DEBUG: pifpaf.drivers: gnocchi-metricd[28413] output: 2016-08-03 00:51:50.229 28444 ERROR cotyledon     self.terminate()
DEBUG: pifpaf.drivers: gnocchi-metricd[28413] output: 2016-08-03 00:51:50.229 28444 ERROR cotyledon   File "/workspace/sileht/os_dev/stack/python-gnocchiclient/.tox/py27/local/lib/python2.7/site-packages/gnocchi/cli.py", line 127, in terminate
DEBUG: pifpaf.drivers: gnocchi-metricd[28413] output: 2016-08-03 00:51:50.229 28444 ERROR cotyledon     self.close_services()
DEBUG: pifpaf.drivers: gnocchi-metricd[28413] output: 2016-08-03 00:51:50.229 28444 ERROR cotyledon   File "/workspace/sileht/os_dev/stack/python-gnocchiclient/.tox/py27/local/lib/python2.7/site-packages/gnocchi/cli.py", line 314, in close_services
DEBUG: pifpaf.drivers: gnocchi-metricd[28413] output: 2016-08-03 00:51:50.229 28444 ERROR cotyledon     self.queue.close()
DEBUG: pifpaf.drivers: gnocchi-metricd[28413] output: 2016-08-03 00:51:50.229 28444 ERROR cotyledon AttributeError: 'AutoProxy[Queue]' object has no attribute 'close'
DEBUG: pifpaf.drivers: gnocchi-metricd[28413] output: 2016-08-03 00:51:50.229 28444 ERROR cotyledon
DEBUG: pifpaf.drivers: gnocchi-metricd[28413] output: 2016-08-03 00:51:50.229 28441 ERROR cotyledon [-] Unhandled exception
DEBUG: pifpaf.drivers: gnocchi-metricd[28413] output: 2016-08-03 00:51:50.229 28441 ERROR cotyledon Traceback (most recent call last):
DEBUG: pifpaf.drivers: gnocchi-metricd[28413] output: 2016-08-03 00:51:50.229 28441 ERROR cotyledon   File "/workspace/sileht/os_dev/stack/python-gnocchiclient/.tox/py27/local/lib/python2.7/site-packages/cotyledon/__init__.py", line 52, in _exit_on_exception
DEBUG: pifpaf.drivers: gnocchi-metricd[28413] output: 2016-08-03 00:51:50.229 28441 ERROR cotyledon     yield
DEBUG: pifpaf.drivers: gnocchi-metricd[28413] output: 2016-08-03 00:51:50.229 28441 ERROR cotyledon   File "/workspace/sileht/os_dev/stack/python-gnocchiclient/.tox/py27/local/lib/python2.7/site-packages/cotyledon/__init__.py", line 141, in _clean_exit
DEBUG: pifpaf.drivers: gnocchi-metricd[28413] output: 2016-08-03 00:51:50.229 28441 ERROR cotyledon     self.terminate()
DEBUG: pifpaf.drivers: gnocchi-metricd[28413] output: 2016-08-03 00:51:50.229 28441 ERROR cotyledon   File "/workspace/sileht/os_dev/stack/python-gnocchiclient/.tox/py27/local/lib/python2.7/site-packages/gnocchi/cli.py", line 127, in terminate
DEBUG: pifpaf.drivers: gnocchi-metricd[28413] output: 2016-08-03 00:51:50.229 28441 ERROR cotyledon     self.close_services()
DEBUG: pifpaf.drivers: gnocchi-metricd[28413] output: 2016-08-03 00:51:50.229 28441 ERROR cotyledon   File "/workspace/sileht/os_dev/stack/python-gnocchiclient/.tox/py27/local/lib/python2.7/site-packages/gnocchi/cli.py", line 314, in close_services
DEBUG: pifpaf.drivers: gnocchi-metricd[28413] output: 2016-08-03 00:51:50.229 28441 ERROR cotyledon     self.queue.close()
DEBUG: pifpaf.drivers: gnocchi-metricd[28413] output: 2016-08-03 00:51:50.229 28441 ERROR cotyledon AttributeError: 'AutoProxy[Queue]' object has no attribute 'close'
DEBUG: pifpaf.drivers: gnocchi-metricd[28413] output: 2016-08-03 00:51:50.229 28441 ERROR cotyledon
DEBUG: pifpaf.drivers: gnocchi-metricd[28413] output: 
DEBUG: pifpaf.drivers: executing: ['/usr/lib/postgresql/9.4/bin/pg_ctl', '-w', 'stop']
ERROR: pifpaf: Traceback (most recent call last):
  File "/workspace/sileht/os_dev/stack/python-gnocchiclient/.tox/py27/local/lib/python2.7/site-packages/fixtures/fixture.py", line 197, in setUp
    self._setUp()
  File "/workspace/sileht/os_dev/stack/python-gnocchiclient/.tox/py27/local/lib/python2.7/site-packages/pifpaf/drivers/gnocchi.py", line 76, in _setUp
    wait_for_line="Metricd reporting")
  File "/workspace/sileht/os_dev/stack/python-gnocchiclient/.tox/py27/local/lib/python2.7/site-packages/pifpaf/drivers/__init__.py", line 135, in _exec
    % (wait_for_line, b"".join(lines)))
RuntimeError: Program did not print: `Metricd reporting'
